### PR TITLE
Mechanik20051988/gh 3633 tuple size increasing

### PR DESCRIPTION
**Pull request by EvgenyMekhanik**
small: implement new allocator strategy
    
    In previous version small allocate memory from most
    appropriate pool. This strategy has one significant
    disadvantage: during memory allocation for tuples with highly
    distributed sizes, we allocate an entire whole slab for
    one or two objects, moreover, when they are further deleted,
    slab is not released (see spare slab in mempool).
    Now we divide mempools with the same slab size into 32 or less
    parts. First, we allocate memory from mempool with the largest
    size, then when memory waste for a certain size of objects
    as a result of a non-optimal pool selection became larger
    then slab size, we start allocating memory for them from the most
    suitable mempool. At the same time, for other objects, we can use
    both of these mempools, in case if new mempool has a larger objsize.
    With this strategy, we avoid losing memory by allocating an entire
    slab for one or two objects.

Perf test results:

[perf.txt](https://github.com/tarantool/small/files/6494644/perf.txt)

### CODE REVIEW

**Review by Korablev77 — COMMENTED** [review]
Overall comments regarding this refactoring:
- Would be nice if before launching test in debug mode warning is appeared saying that it may take SIGNIFICANTLY (up to 100 times) more time than in release mode;
- make fractional rps to make results more precise;
- Display legend before launching test: passed/available test modes; what does mrps mean;
- I'd split number of allocations/deallocations (remaining the total number of operations);

File: perf/small_alloc_perf.c [hunk 1]
```diff
@@ -8,13 +8,40 @@
...
+/**
+ * True in the case where we don't need to test performance
+ * of memory allocation for large size objects.
+ */
+static bool no_large_test = false;
+/**
+ * True in the case where we don't need to
+ * test delayed free mode.
+ */
+static bool no_delayed_free_mode = false;
```
Comment (Korablev77) [inline perf/small_alloc_perf.c:44]: What's delayed free mode?

File: perf/small_alloc_perf.c [hunk 2]
```diff
@@ -100,46 +139,79 @@ small_is_unused(void)
...
-			small_alloc_setopt(&alloc,
-					   SMALL_DELAYED_FREE_MODE, true);
-			break;
-		default:
-			break;
+		if (!no_delayed_free_mode) {
+			int mode = i % 3;
+			switch (mode) {
+			case 1:
+				small_alloc_setopt(&alloc,
```
Comment (Korablev77) [inline perf/small_alloc_perf.c:168]: Don't get this part. If we have no_delayed_free_mode, why do we have to set it anyway once per three iterations?

Comment (Korablev77) [reply perf/small_alloc_perf.c:168]: If it is still required, could you please use enum for mode values (to make it a bit comprehensible).

File: perf/small_alloc_perf.c [hunk 3]
```diff
@@ -333,6 +515,26 @@ int main(int argc, char* argv[])
 	if (argc == 2 && !strcmp(argv[1], "-h")) //human clear output
 		human = true;
 
+	for (int i = 1; i < argc; i++) {
+		if (!strcmp(argv[i], "-h")) {
```
Comment (Korablev77) [inline perf/small_alloc_perf.c:519]: -h as a rule means --help and provides information about available launch keys and assists user in launching program..
I suggest to rework -h flag. For changing results format we can use another flag.

File: perf/small_alloc_perf.c [hunk 4]
```diff
@@ -181,103 +253,206 @@ print_json_test_finish(const char * finish)
...
-			"_________________________________|\n");
-	} else {
-		print_json_test_header("random");
+	if (!no_basic_random_test) {
+		/*
+		 * Memory allocation/deallocation performance test for random size objects.
+		 */
+		if(human) {
+			fprintf(stderr, "|              SMALL RANDOM "
+				"ALLOCATION RESULT TABLE                  |\n");
```
Comment (Korablev77) [inline perf/small_alloc_perf.c:283]: These prints repeat the same pattern in different places. Mb it is worth extracting it to the function?

**Review by Korablev77 — COMMENTED** [review]
How does this change affect performance?

File: small/mempool.h [hunk 5]
```diff
@@ -278,6 +288,28 @@ mempool_free(struct mempool *pool, void *ptr)
 	mslab_free(pool, slab, ptr);
 }
 
+/**
+ * Free a single object.
```
Comment (Korablev77) [inline small/mempool.h:292]: What's the difference between this function and original one (mempool_free)?

File: small/small.c [hunk 6]
```diff
@@ -202,15 +299,15 @@ void
 smfree(struct small_alloc *alloc, void *ptr, size_t size)
 {
 	struct mempool *pool = mempool_find(alloc, size);
-	if (pool == NULL) {
+	if (unlikely(pool == NULL)) {
```
Comment (alyapunov) [inline small/small.c:302]: That's an optimization and is not relevant to current commit. Please move it to separate commit.

File: small/small.c [hunk 7]
```diff
@@ -223,7 +320,7 @@ smfree_delayed(struct small_alloc *alloc, void *ptr, size_t size)
 {
 	if (alloc->free_mode == SMALL_DELAYED_FREE && ptr) {
 		struct mempool *pool = mempool_find(alloc, size);
-		if (pool == NULL) {
+		if (unlikely(pool == NULL)) {
```
Comment (alyapunov) [inline small/small.c:323]: That's an optimization and is not relevant to current commit. Please move it to separate commit.

File: small/small.c [hunk 8]
```diff
@@ -162,27 +258,28 @@ smalloc(struct small_alloc *alloc, size_t size)
 {
 	small_collect_garbage(alloc);
 
-	struct factor_pool *upper_bound = factor_pool_search(alloc, size);
-	if (upper_bound == NULL) {
+	struct factor_pool *factor_pool = factor_pool_search(alloc, size);
+	if (unlikely(factor_pool == NULL)) {
```
Comment (alyapunov) [inline small/small.c:262]: That's an optimization and is not relevant to current commit. Please move it to separate commit.

File: small/small.c [hunk 9]
```diff
@@ -162,27 +258,28 @@ smalloc(struct small_alloc *alloc, size_t size)
 {
 	small_collect_garbage(alloc);
 
-	struct factor_pool *upper_bound = factor_pool_search(alloc, size);
```
Comment (alyapunov) [inline small/small.c:165]: Rename `upper_bound` -> `factor_pool` is a candidate for a journey to another commit.

File: small/small.c [hunk 10]
```diff
@@ -32,23 +32,90 @@
...
+		pool->used_pool->pool.objsize - pool->pool.objsize;
+	if (unlikely(pool->pool.larger_pool_waste_cur >=
+		     pool->pool.larger_pool_waste_max))
+		factor_pool_update_used_pool(pool);
+
+	return pool->used_pool;
+}
+
+static inline struct factor_pool*
+caclulate_last_pool_for_mask(struct factor_pool *cur, struct factor_pool *last)
```
Comment (alyapunov) [inline small/small.c:101]: Misprint: caclulate.

File: small/small.c [hunk 11]
```diff
@@ -32,23 +32,90 @@
...
+{
+	if (unlikely(size > alloc->objsize_max))
+		return NULL;
+	unsigned cls = small_class_calc_offset_by_size(&alloc->small_class,
+						       size);
+	struct factor_pool *pool = &alloc->factor_pool_cache[cls];
+	if (likely(pool->used_pool == pool))
+		return pool;
+
+	pool->pool.larger_pool_waste_cur +=
```
Comment (alyapunov) [inline small/small.c:91]: I didn't understand the idea. There must be comments or something.

**Comment by alyapunov** [comment]
The patch is very hard to understand (and support).
I think we should start with improvements:
1. Please move non-logic change to another commit.
2. Each allocator - clab_cache, mempool and smalloc - must be self-contained. 
3. Not obvious code (there's lots of it here) must be well-commented.
4. We discussed term `mempool group`. Maybe using it will help to make the logic more understandable.
5. smalloc and smfree looks quite different now. I think there must be a cleaner way.

I'll think a little and the we should discuss it offline.

**Review by alyapunov — CHANGES_REQUESTED** [review]

**Comment by Korablev77** [comment]
I'm a bit late with my own comments, but I share almost all @alyapunov  concerns.

File: perf/small_alloc_perf.c [hunk 12]
```diff
@@ -8,13 +8,40 @@
...
+unsigned int large_iterations_count = 1000;
 const size_t SLAB_SIZE_MIN = 4 * 1024 * 1024;
 const size_t SLAB_SIZE_MAX = 16 * 1024 * 1024;
 static const unsigned long long NANOSEC_PER_SEC  = 1000000000;
-static const unsigned long long NANOSEC_PER_MSEC  = 1000000;
+/**
+ * True in the case where we don't need to test performance
+ * of memory allocation for random size objects.
+ */
+static bool no_basic_random_test = false;
```
[the remark this answers is not in the document]
Comment (mary3000) [reply perf/small_alloc_perf.c:24]: I skimmed through the PR, there are many double negations like that: `!no_basic_random_test` - which is hard to read

File: perf/small_alloc_perf.c [hunk 13]
```diff
@@ -51,13 +81,14 @@ free_checked(int *ptr)
...
-calculate_pow_factor(int size_max, int pow_max, int start)
+calculate_pow_factor(const int size_max, const int pow_max, const int start)
 {
 	return exp(log((double)size_max / start) / pow_max);
 }
 
 static inline void *
-alloc_checked(int pos, int size_min, int size_max, int rnd, double pow_factor)
+alloc_checked(const int pos, const int size_min, const int size_max,
+	      const int rnd, const double pow_factor)
```
[the remark this answers is not in the document]
Comment (Korablev77) [reply perf/small_alloc_perf.c:91]: Also as a rule we don't make arguments constant (especially when it comes for copied objects). So I wouldn't change 'constantness' of arguments.

**Review by alyapunov — CHANGES_REQUESTED** [review]
In test all sizes should be of size_t type (not unsigned or signed int). Actually we can use smaller types but if we can explain profit (like significant memory saving).

Please replace std::pair with struct allocation { void *ptr; size_t size; }; There's no reason to be nameless here.

Just thoughts - do we build perf test on our CI? It seems we should.

File: small/mempool.c [hunk 14]
```diff
@@ -145,11 +145,11 @@ mslab_free(struct mempool *pool, struct mslab *slab, void *ptr)
 			slab_put_with_order(pool->cache, &pool->spare->slab);
 			pool->spare = slab;
 		 } else if (pool->spare) {
-			 slab_list_del(&pool->slabs, &slab->slab,
+			slab_list_del(&pool->slabs, &slab->slab,
 				       next_in_list);
```
Comment (Korablev77) [inline small/mempool.c:151]: Align next_in_list under &pool->slabs. Also you can remove dots before }:
```
diff --git a/small/mempool.c b/small/mempool.c
index edb8cac..5252767 100644
--- a/small/mempool.c
+++ b/small/mempool.c
@@ -146,13 +146,13 @@ mslab_free(struct mempool *pool, struct mslab *slab, void *ptr)
                                      next_in_list);
                        slab_put_with_order(pool->cache, &pool->spare->slab);
                        pool->spare = slab;
-                } else if (pool->spare) {
+               } else if (pool->spare) {
                        slab_list_del(&pool->slabs, &slab->slab,
-                                      next_in_list);
+                                     next_in_list);
                        slab_put_with_order(pool->cache, &slab->slab);
-                } else {
+               } else {
                        pool->spare = slab;
-                }
+               }
        }
```

File: small/small.h [hunk 15]
```diff
@@ -81,6 +81,34 @@ struct small_mempool
...
 	 */
 	size_t objsize_min;
+	/**
+	 * Mask for this pool, which shows all pools in the group with an
+	 * object size greater than or equal to the object size of this pool,
+	 * from which there was at least one allocation.
+	 */
+	uint32_t used_pool_mask;
+	/**
+	 * Index of this factor pool in factor pool group with same slab
```
Comment (Korablev77) [inline small/small.h:91]: There's no more factor pools as I understand.

File: small/small.c [hunk 16]
```diff
@@ -61,7 +179,32 @@ small_mempool_create(struct small_alloc *alloc)
...
+		 * mempools count with same slab size. There can be no more
+		 * than 32 pools in one group.
+		 */
+		if (slab_order_next != slab_order_cur) {
+			slab_order_cur = slab_order_next;
+			small_mempool_create_groups(cur_order_pool, pool - 1);
+			cur_order_pool = pool;
+		}
+
+		if (objsize == alloc->objsize_max)
```
Comment (Korablev77) [inline small/small.c:251]: Don't get this edge case. Could you please add an explanation?

File: small/small.h (hunk 15 above)
Comment (EvgenyMekhanik) [reply small/small.h:91]: With a low allocation factor, there can be more than 32 pools with the same slab size. We limit this number to 32 pools and, if we have more than 32 pools with same slab size, we make two or more groups of pools with the same slab size

File: small/mempool.h [hunk 17]
```diff
@@ -261,23 +274,37 @@ void
...
- * @pre the object is allocated in this pool.
+ * Helper function for quick free up memory. In case we know
+ * slab we don't need to find it from ptr. Used in case when
+ * mempool is a part of small_alloc. We find slab from ptr in
+ * smfree function and we don't need to search for it again to
+ * free up memory.
  */
 static inline void
-mempool_free(struct mempool *pool, void *ptr)
+mempool_free_slab(struct mempool *pool, struct mslab *slab, void *ptr)
```
Comment (Korablev77) [inline small/mempool.h:284]: Hm, mslab now has pointer to its mempool, why do we need then pass mempool to free slab?
Anyway assert(slab->mempool == pool) would be suitable

File: small/small.h [hunk 18]
```diff
@@ -80,6 +82,37 @@ struct small_mempool {
 	 * pool.
 	 */
 	size_t objsize_min;
+	/** Small mempool group that this pool belongs to. */
+	struct small_mempool_group *group;
+	/** Currently used pool for memory allocation. */
+	struct small_mempool *used_pool;
+	/** Mask of appropriate pools. */
```
Comment (Korablev77) [inline small/small.h:89]: What's an appropriate pool?

**Comment by EvgenyMekhanik** [comment]
New pool request in other branch np/mempool-groups
